cpp/include/raft/core/mdarray.hpp

```diff
@@ -37,6 +37,96 @@ using layout_c_contiguous = detail::stdex::layout_right;
  */
 using layout_f_contiguous = detail::stdex::layout_left;
 
+using layout_stride = detail::stdex::layout_stride;
+
+/**
+ * @brief row major padded layout mapping for mdarray and mdspan.
+ */
+template <class Extents, size_t ElementAlignment>
+class layout_padded_right : public layout_stride::mapping<Extents> {
+ private:
+  using stride_storage_type = std::array<size_t, Extents::rank()>;
+
+  stride_storage_type generate_row_major_strides_with_padding(Extents const& __exts)
+  {
+    auto alignment              = std::max<size_t>(ElementAlignment, 1);
+    stride_storage_type strides = std::array<size_t, Extents::rank()>{};
+    size_t stride               = 1;
+    for (size_t r = __exts.rank() - 1; r > 0; r--) {
+      strides[r] = stride;
+      if (stride == 1) {
+        stride *=
+          std::max<size_t>(alignment, (__exts.extent(r) + alignment - 1) / alignment * alignment);
```

Comment:
A nitpick:
if ElementAlignment is a power of two, you can use `Pow2<ElementAlignment>::roundUp( __exts.extent(r))` from [pow2_utils.cuh](https://github.com/rapidsai/raft/blob/6e8e41db27793e45ecc41360f1c82c0a0f1e1de0/cpp/include/raft/pow2_utils.cuh#L137)